src/planner/merge_node.go

```diff
@@ -184,15 +198,98 @@ func (m *MergeNode) calcRoute() (SelectNode, error) {
 // pushSelectExprs used to push the select fields.
 func (m *MergeNode) pushSelectExprs(fields, groups []selectTuple, sel *sqlparser.Select, aggTyp aggrType) error {
 	node := m.Sel.(*sqlparser.Select)
-	node.SelectExprs = sel.SelectExprs
-	node.GroupBy = sel.GroupBy
+	if node.SelectExprs == nil {
+		node.SelectExprs = sel.SelectExprs
+	}
+	if m.fields == nil {
+		node.GroupBy = sel.GroupBy
+	}
 	node.Distinct = sel.Distinct
-	m.fields = fields
+	if m.fields == nil {
+		m.fields = fields
+	} else {
+		for _, subPlan := range m.children.children {
+			if aggPlan, ok := subPlan.(*AggregatePlan); ok {
+				if len(aggPlan.normalAggrs) > 0 {
+					return errors.Errorf("unsupported: aggregation.function.in.subquery")
+				}
+				return errors.Errorf("unsupported: group.by.in.subquery")
+			}
+		}
+		if aggTyp == canPush {
+			aggTyp = notPush
+		}
+
+		//rebuild the fields
+		node.SelectExprs = nil
+		unusedFields := make([]selectTuple, len(m.fields))
+		dupFields := make([]selectTuple, len(m.fields))
+		copy(unusedFields, m.fields)
+		copy(dupFields, m.fields)
+		var newFields []selectTuple
+
+		for _, field := range fields {
+			if field.field == "*" {
+				for _, f := range m.fields {
+					newFields = append(newFields, f)
+					node.SelectExprs = append(node.SelectExprs, f.expr)
+				}
+			} else {
+				if field.aggrFuc == "" {
+					f, err := getMatchedField(field, m.fields)
+					if err != nil {
+						return err
+					}
+					newFields = append(newFields, f)
+					node.SelectExprs = append(node.SelectExprs, f.expr)
+				} else {
+					if field.aggrField == "*" {
+						newFields = append(newFields, field)
+						node.SelectExprs = append(node.SelectExprs, field.expr)
+						continue
+					}
+					s := strings.Split(field.aggrField, ".")
+					field.field = s[len(s)-1]
+					f, err := getMatchedField(field, m.fields)
+					if err != nil {
+						return err
+					}
+					f.aggrFuc = field.aggrFuc
+					field.expr.(*sqlparser.AliasedExpr).Expr.(*sqlparser.FuncExpr).Exprs[0] = f.expr.(*sqlparser.AliasedExpr)
+					f.expr = field.expr
+					f.aggrField = field.field
+					if len(f.referTables) > 0 {
+						f.aggrField = f.referTables[0] + "." + f.field
+					}
+					f.alias = field.alias
+					f.field = f.aggrFuc + "(" + f.aggrField + ")"
+
+					newFields = append(newFields, f)
+					node.SelectExprs = append(node.SelectExprs, f.expr)
+				}
+			}
+		}
+		m.fields = newFields
+	}
 
 	if len(sel.GroupBy) > 0 {
 		// group by implicitly contains order by.
 		if len(sel.OrderBy) == 0 {
 			for _, by := range sel.GroupBy {
+				e, ok := by.(*sqlparser.ColName)
```

Comment:
In `select_plan.go`, we have `checkgroupby` yet, why check group by agent.
Group by depends on the same layer of fields, and does not need to pay attention to subquery fields.

Reply:
Such as :
select  p.a from (select  t1.a  from t1) p group by p.a ;
There is a need to transfrom "group by p.a" into real table field .

Reply:
In fact the code is useful only if exists a from subquery. In other cases,the code is unnecessary. So how about add a `if` statement  to the front of the code.

Reply:
There is a ```if``` statement after this statement，the  ```if``` statement relys on variable ```e```.

Reply:
`select b from t3 group by b;` also can enter the logic.